File: yarn-project/acir-simulator/src/avm/opcodes/opcodes.ts
/**
 * All AVM opcodes. (Keep in sync with cpp counterpart code AvmMini_opcode.hpp).
 * Source: https://yp-aztec.netlify.app/docs/public-vm/instruction-set
 */
export enum Opcode {
  // Compute
  // Compute - Arithmetic
  ADD = 0x00,
  SUB = 0x01,
  MUL = 0x02,
  DIV = 0x03,
  // Compute - Comparators
  EQ = 0x04,
  LT = 0x05,
  LTE = 0x06,
  // Compute - Bitwise
  AND = 0x07,
  OR = 0x08,
  XOR = 0x09,
  NOT = 0x0a,
  SHL = 0x0b,
  SHR = 0x0c,
  // Compute - Type Conversions
  CAST = 0x0d,

  // Execution Environment
  ADDRESS = 0x0e,
  STORAGEADDRESS = 0x0f,
  ORIGIN = 0x10,
  SENDER = 0x11,
  PORTAL = 0x12,
  FEEPERL1GAS = 0x13,
  FEEPERL2GAS = 0x14,
  FEEPERDAGAS = 0x15,
  CONTRACTCALLDEPTH = 0x16,
  // Execution Environment - Globals
  CHAINID = 0x17,
  VERSION = 0x18,
  BLOCKNUMBER = 0x19,
  TIMESTAMP = 0x1a,
  COINBASE = 0x1b,
  BLOCKL1GASLIMIT = 0x1c,
  BLOCKL2GASLIMIT = 0x1d,
  BLOCKDAGASLIMIT = 0x1e,
  // Execution Environment - Calldata
  CALLDATACOPY = 0x1f,

  // Machine State
  // Machine State - Gas
  L1GASLEFT = 0x20,
  L2GASLEFT = 0x21,
  DAGASLEFT = 0x22,
  // Machine State - Internal Control Flow
  JUMP = 0x23,
  JUMPI = 0x24,
  INTERNALCALL = 0x25,
  INTERNALRETURN = 0x26,
  // Machine State - Memory
  SET = 0x27,
  MOV = 0x28,
  CMOV = 0x29,

  // World State
  BLOCKHEADERBYNUMBER = 0x2a,
  SLOAD = 0x2b, // Public Storage
  SSTORE = 0x2c, // Public Storage
  READL1TOL2MSG = 0x2d, // Messages
  SENDL2TOL1MSG = 0x2e, // Messages
  EMITNOTEHASH = 0x2f, // Notes & Nullifiers
  EMITNULLIFIER = 0x30, // Notes & Nullifiers

  // Accrued Substate
  EMITUNENCRYPTEDLOG = 0x31,

  // Control Flow - Contract Calls
  CALL = 0x32,
  STATICCALL = 0x33,
  RETURN = 0x34,
  REVERT = 0x35,

  // Gadgets
  KECCAK = 0x36,
  POSEIDON = 0x37,

  // Add new opcodes before this
  TOTAL_OPCODES_NUMBER,
}

File: avm-transpiler/src/opcodes.rs
/// All AVM opcodes
/// Keep updated with TS and yellow paper!
#[derive(PartialEq, Copy, Clone, Debug)]
pub enum AvmOpcode {
    // Compute
    ADD,
    SUB,
    MUL,
    DIV,
    FDIV,
    EQ,
    LT,
    LTE,
    AND,
    OR,
    XOR,
    NOT,
    SHL,
    SHR,
    CAST,
    // Execution environment
    ADDRESS,
    STORAGEADDRESS,
    ORIGIN,
    SENDER,
    PORTAL,
    FEEPERL1GAS,
    FEEPERL2GAS,
    FEEPERDAGAS,
    CONTRACTCALLDEPTH,
    CHAINID,
    VERSION,
    BLOCKNUMBER,
    TIMESTAMP,
    COINBASE,
    BLOCKL1GASLIMIT,
    BLOCKL2GASLIMIT,
    BLOCKDAGASLIMIT,
    CALLDATACOPY,
    // Gas
    L1GASLEFT,
    L2GASLEFT,
    DAGASLEFT,
    // Control flow
    JUMP,
    JUMPI,
    INTERNALCALL,
    INTERNALRETURN,
    // Memory
    SET,
    MOV,
    CMOV,
    // World state
    SLOAD,
    SSTORE,
    NOTEHASHEXISTS,
    EMITNOTEHASH,
    NULLIFIEREXISTS,
    EMITNULLIFIER,
    L1TOL2MSGEXISTS,
    HEADERMEMBER,
    GETCONTRACTINSTANCE,
    EMITUNENCRYPTEDLOG,
    SENDL2TOL1MSG,
    // External calls
    CALL,
    STATICCALL,
    DELEGATECALL,
    RETURN,
    REVERT,
    // Gadgets
    KECCAK,
    POSEIDON,
    SHA256,   // temp - may be removed, but alot of contracts rely on it
    PEDERSEN, // temp - may be removed, but alot of contracts rely on it
}

impl AvmOpcode {
    pub fn name(&self) -> &'static str {
        match self {
            // Compute
            // Compute - Arithmetic
            AvmOpcode::ADD => "ADD",
            AvmOpcode::SUB => "SUB",
            AvmOpcode::MUL => "MUL",
            AvmOpcode::DIV => "DIV",
            AvmOpcode::FDIV => "FDIV",
            // Compute - Comparators
            AvmOpcode::EQ => "EQ",
            AvmOpcode::LT => "LT",
            AvmOpcode::LTE => "LTE",
            // Compute - Bitwise
            AvmOpcode::AND => "AND",
            AvmOpcode::OR => "OR",
            AvmOpcode::XOR => "XOR",
            AvmOpcode::NOT => "NOT",
            AvmOpcode::SHL => "SHL",
            AvmOpcode::SHR => "SHR",
            // Compute - Type Conversions
            AvmOpcode::CAST => "CAST",

            // Execution Environment
            AvmOpcode::ADDRESS => "ADDRESS",
            AvmOpcode::STORAGEADDRESS => "STORAGEADDRESS",
            AvmOpcode::ORIGIN => "ORIGIN",
            AvmOpcode::SENDER => "SENDER",
            AvmOpcode::PORTAL => "PORTAL",
            AvmOpcode::FEEPERL1GAS => "FEEPERL1GAS",
            AvmOpcode::FEEPERL2GAS => "FEEPERL2GAS",
            AvmOpcode::FEEPERDAGAS => "FEEPERDAGAS",
            AvmOpcode::CONTRACTCALLDEPTH => "CONTRACTCALLDEPTH",
            // Execution Environment - Globals
            AvmOpcode::CHAINID => "CHAINID",
            AvmOpcode::VERSION => "VERSION",
            AvmOpcode::BLOCKNUMBER => "BLOCKNUMBER",
            AvmOpcode::TIMESTAMP => "TIMESTAMP",
            AvmOpcode::COINBASE => "COINBASE",
            AvmOpcode::BLOCKL1GASLIMIT => "BLOCKL1GASLIMIT",
            AvmOpcode::BLOCKL2GASLIMIT => "BLOCKL2GASLIMIT",
            AvmOpcode::BLOCKDAGASLIMIT => "BLOCKDAGASLIMIT",
            // Execution Environment - Calldata
            AvmOpcode::CALLDATACOPY => "CALLDATACOPY",

            // Machine State
            // Machine State - Gas
            AvmOpcode::L1GASLEFT => "L1GASLEFT",
            AvmOpcode::L2GASLEFT => "L2GASLEFT",
            AvmOpcode::DAGASLEFT => "DAGASLEFT",
            // Machine State - Internal Control Flow
            AvmOpcode::JUMP => "JUMP",
            AvmOpcode::JUMPI => "JUMPI",
            AvmOpcode::INTERNALCALL => "INTERNALCALL",
            AvmOpcode::INTERNALRETURN => "INTERNALRETURN",
            // Machine State - Memory
            AvmOpcode::SET => "SET",
            AvmOpcode::MOV => "MOV",
            AvmOpcode::CMOV => "CMOV",

            // World State
            AvmOpcode::SLOAD => "SLOAD",   // Public Storage
            AvmOpcode::SSTORE => "SSTORE", // Public Storage
            AvmOpcode::NOTEHASHEXISTS => "NOTEHASHEXISTS", // Notes & Nullifiers
            AvmOpcode::EMITNOTEHASH => "EMITNOTEHASH", // Notes & Nullifiers
            AvmOpcode::NULLIFIEREXISTS => "NULLIFIEREXISTS", // Notes & Nullifiers
            AvmOpcode::EMITNULLIFIER => "EMITNULLIFIER", // Notes & Nullifiers
            AvmOpcode::L1TOL2MSGEXISTS => "L1TOL2MSGEXISTS", // Messages
            AvmOpcode::HEADERMEMBER => "HEADERMEMBER", // Archive tree & Headers

            // Accrued Substate
            AvmOpcode::EMITUNENCRYPTEDLOG => "EMITUNENCRYPTEDLOG",
            AvmOpcode::SENDL2TOL1MSG => "SENDL2TOL1MSG",
            AvmOpcode::GETCONTRACTINSTANCE => "GETCONTRACTINSTANCE",

            // Control Flow - Contract Calls
            AvmOpcode::CALL => "CALL",
            AvmOpcode::STATICCALL => "STATICCALL",
            AvmOpcode::DELEGATECALL => "DELEGATECALL",
            AvmOpcode::RETURN => "RETURN",
            AvmOpcode::REVERT => "REVERT",

            // Gadgets
            AvmOpcode::KECCAK => "KECCAK",
            AvmOpcode::POSEIDON => "POSEIDON",
            AvmOpcode::SHA256 => "SHA256 ",
            AvmOpcode::PEDERSEN => "PEDERSEN",
        }
    }
}
